src/lib.rs

```diff
@@ -3,58 +3,121 @@ extern crate std;
 
 use alloc::format;
 use alloc::string::{String, ToString};
-use std::io::{BufReader, Read};
+use std::io::{BufReader, Read, BufRead};
+use sha2::{Sha256, Digest};
 
-pub fn generate_sha1_git_oid(x: &[u8]) -> String {
-    let prefix = format!("blob {}\0", x.len());
-    let mut hasher = sha1_smol::Sha1::new();
+use base64ct::{Base64, Encoding};
 
-    hasher.update(prefix.as_bytes());
-    hasher.update(x);
+enum HashAlgorithm {
+    SHA1,
+    SHA256
+}
 
-    hasher.digest().to_string()
+struct GitOid {
+    hash_algorithm: HashAlgorithm,
 }
 
-pub fn generate_sha1_git_oid_from_buffer<R>(
-    mut reader: BufReader<R>,
-    expected_length: usize,
-) -> String
-where
-    BufReader<R>: std::io::Read,
-{
-    let prefix = format!("blob {}\0", expected_length);
-    let mut hasher = sha1_smol::Sha1::new();
-
-    hasher.update(prefix.as_bytes());
-
-    let mut buf = [0; 4096]; // linux default page size is 4096
-    let mut amount_read = 0;
-    loop {
-        let y = reader.read(&mut buf);
-        match y {
-            Ok(0) => {
-                break;
-            }
-            Ok(size) => {
-                hasher.update(&buf[..size]);
-                amount_read = amount_read + size;
-            }
-            Err(_) => {
-                break;
+impl GitOid {
+    pub fn generate_git_oid(&self, content: &[u8]) -> String {
+        let prefix = format!("blob {}\0", content.len());
+
+        match self.hash_algorithm {
+            HashAlgorithm::SHA1 => {
+                let mut hasher = sha1_smol::Sha1::new();
+
+                hasher.update(prefix.as_bytes());
+                hasher.update(content);
+
+                hasher.digest().to_string()
+            },
+            HashAlgorithm::SHA256 => {
+                let mut hasher = Sha256::new();
+
+                hasher.update(prefix.as_bytes());
+                hasher.update(content);
+
+                let hash = hasher.finalize();
+                let hash_string = Base64::encode_string(&hash);
+
+                return Base64::encode_string(&hash)
             }
         }
     }
 
-    hasher.digest().to_string()
+    pub fn generate_git_oid_from_buffer<R>(
+        &self,
+        mut reader: BufReader<R>,
+        expected_length: usize,
+    ) -> String
+    where
+        BufReader<R>: std::io::Read,
+    {
+        let prefix = format!("blob {}\0", expected_length);
+
+        let mut buf = [0; 4096]; // linux default page size is 4096
+        let mut amount_read = 0;
+
+        match self.hash_algorithm {
+            HashAlgorithm::SHA1 => {
+                let mut hasher = sha1_smol::Sha1::new();
+
+                hasher.update(prefix.as_bytes());
```

Comment:
I considered consolidating the following loop into a separate function that is called for both hash algorithms. However, the type of hash is different (for sha1_smol it's sha1, for Sha256 it's Core Wrapper). We could use a generic type, but that feels like overkill at this point (particularly if we move to a different sha1 library which might return the same type).